docs/design/2018-07-22-enhance-propagations.md

```diff
@@ -0,0 +1,189 @@
+# Proposal: Enhance constraint propagation in TiDB logical plan
+
+- Author(s):     [@bb7133](https://github.com/bb7133), [@zz-jason](https://github.com/zz-jason) 
+- Last updated:  2018-07-22
+- Discussion at: https://docs.google.com/document/d/1G3wVBaiza9GI5q9nwHLCB2I7r1HUbg6RpTNFpQWTFhQ/edit#
+
+## Abstract
+
+This proposal tries to illustrate some rules that can be added to current constraint/filter propgation optimizations in TiDB logcial plan.
+
+## Background
+
+For now, most of the constraint propagation work in TiDB is done by `propagateConstantSolver`, it does:
+
+1. Find `column = constant` expression and substitue the constant for column, as well as try to fold the substituted constant expression if possible, for example:
+
+ Given `a = b and a = 2 and b = 3`, it becomes `2 = 3` after substitution and lead to a final `false` constant.
+
+2. Find `column A = column B` expression(which happens in `join` statements mostly) and propagate expressions like `column op constant`(as well as `constant op column`) based on the equliaty relation, the supported operators are:
+
+ * `ast.LT('<')`
+ * `ast.GT('>')`
+ * `ast.LE('<=')`
+ * `ast.GE('>=')`
+ * `ast.NE('!=')`
+
+The `propagateConstantSolver` makes more detailed/explicit filters/constraints, which can be used within other optimization rule. For example, in predicate-pushdown optimization, it generates more predicates that can be pushed closer to the data source(TiKV), and thus reduce the amount of data in the whole data path.
+
+We can further do the optimization by introduce more rules and infer/propagate more constraints from the existings ones, which helps us building better logical plan.
+
+## Proposal
+
+Here are proposed rules we can consider:
+
+1. Infer more filters/constraints from column equality relation
+
+ We should be able to infer set of data constraints based on column equality, that is, any constraint on `a` now applies to `b` as long as:
+
+ * The constraint is deterministic
+ * The constraint doesn’t have any side effect
+
+ For example, 
+
+ `t1.a = t2.a and t1.a < 5` => `t2.a < 5`
```

Comment:
To simply describe the examples, maybe it's better to `s/t1.a/a/` and `s/t2.a/b/`? By the way, how about using a table to describe the examples, for example:

| origin filters | propagated filters |
|----------------|--------------------|
|t1.a = t2.a and t1.a < 5 | t2.a < 5 |
|t1.a = t2.a and t1.a in (12, 13) and t2.a in (14, 15) | t1.a in (14, 15) and t2.a in (12, 13) |

Reply:
Good idea, thanks!